File: internal/skill/skill.go
package skill

type SkillContext struct {
	SelectedCode string   `json:"selectedCode"`
	FilePath     string   `json:"filePath"`
	FileContent  string   `json:"fileContent"`
	Diagnostics  []string `json:"diagnostics"`
	Language     string   `json:"language"`
	ProjectPath  string   `json:"projectPath"`
	ContextFiles []string `json:"contextFiles,omitempty"`
	UserInput    string   `json:"userInput,omitempty"`
}

type SkillDef struct {
	ID               string   `json:"id"`
	Name             string   `json:"name"`
	Icon             string   `json:"icon"`
	Description      string   `json:"description"`
	Trigger          string   `json:"trigger"`
	PromptTemplate   string   `json:"promptTemplate"`
	ResultType       string   `json:"resultType"`
	AssociatedAgents []string `json:"associatedAgents"`
	Category         string   `json:"category"`
}

type SkillResult struct {
	SkillID    string `json:"skillId"`
	Content    string `json:"content"`
	ResultType string `json:"resultType"`
}

type SkillPipelineStep struct {
	SkillID   string `json:"skillId"`
	Condition string `json:"condition,omitempty"`
	InputFrom string `json:"inputFrom,omitempty"`
	Optional  bool   `json:"optional,omitempty"`
}

type SkillPipeline struct {
	ID          string              `json:"id"`
	Name        string              `json:"name"`
	Description string              `json:"description"`
	Steps       []SkillPipelineStep `json:"steps"`
}

type PipelineStepResult struct {
	StepIndex int          `json:"stepIndex"`
	SkillID   string       `json:"skillId"`
	Result    *SkillResult `json:"result"`
	Skipped   bool         `json:"skipped,omitempty"`
	Error     string       `json:"error,omitempty"`
}

type PipelineExecutionResult struct {
	PipelineID string               `json:"pipelineId"`
	Steps      []PipelineStepResult `json:"steps"`
	Success    bool                 `json:"success"`
	Error      string               `json:"error,omitempty"`
}

File: frontend/src/stores/skill.js
import { writable } from 'svelte/store'
import { EventsOn, EventsOff } from '../../wailsjs/runtime/runtime.js'
import { customModels, resolveModelProvider } from './provider.js'

/** @typedef {{ id: string, name: string, icon: string, description: string, trigger: string, resultType: string, associatedAgents: string[], category: string }} SkillDef */
/** @typedef {{ selectedCode: string, filePath: string, fileContent: string, diagnostics: string[], language: string, projectPath: string }} SkillContext */

export const skills = writable(/** @type {SkillDef[]} */ ([]))
export const executingSkillId = writable(/** @type {string|null} */ (null))
export const skillResult = writable(/** @type {string} */ (''))
export const isSkillExecuting = writable(false)

export async function loadSkills() {
  if (!window.backend?.GetSkills) return
  try {
    /** @type {SkillDef[]} */
    const list = await window.backend.GetSkills()
    skills.set(list || [])
  } catch (/** @type {any} */ e) {
    console.error('Failed to load skills:', e)
  }
}

/**
 * @param {string} skillId
 * @param {SkillContext} context
 * @param {string} providerId
 * @param {string} model
 */
export async function executeSkill(skillId, context, providerId, model) {
  if (!window.backend?.ExecuteSkill) return
  executingSkillId.set(skillId)
  isSkillExecuting.set(true)
  skillResult.set('')

  // Resolve custom model and auto-configure provider before executing skill
  const resolved = resolveModelProvider(model, providerId, customModels)
  if (resolved.apiKey || resolved.endpoint) {
    try {
      await window.backend.SetProviderConfig(resolved.providerId, {
        id: resolved.providerId,
        name: resolved.providerId,
        apiKey: resolved.apiKey || '',
        endpoint: resolved.endpoint || '',
        enabled: true,
        isDefault: false,
      })
    } catch (e) {
      console.error('Failed to set provider config for skill:', e)
    }
  }

  const startEvent = 'skill:stream:start'
  const dataEvent = 'skill:stream:data'
  const doneEvent = 'skill:stream:done'
  const errorEvent = 'skill:stream:error'

  let result = ''

  const offStart = EventsOn(startEvent, (/** @type {string} */ id) => {
    executingSkillId.set(id)
  })

  const offData = EventsOn(dataEvent, (/** @type {string} */ chunk) => {
    result += chunk
    skillResult.set(result)
  })

  const offDone = EventsOn(doneEvent, () => {
    cleanup()
  })

  const offError = EventsOn(errorEvent, (/** @type {string} */ err) => {
    if (result === '') {
      skillResult.set(`Error: ${err}`)
    }
    cleanup()
  })

  function cleanup() {
    offStart()
    offData()
    offDone()
    offError()
    EventsOff(startEvent)
    EventsOff(dataEvent)
    EventsOff(doneEvent)
    EventsOff(errorEvent)
    isSkillExecuting.set(false)
  }

  try {
    await window.backend.ExecuteSkill(skillId, context, resolved.providerId, resolved.model)
  } catch (/** @type {any} */ e) {
    console.error('Skill execution failed:', e)
    skillResult.set(`Error: ${e.message || String(e)}`)
    cleanup()
  }
}

export function clearSkillResult() {
  skillResult.set('')
  executingSkillId.set(null)
}

/**
 * @template T
 * @param {import('svelte/store').Readable<T>} store
 * @returns {T}
 */
function get(store) {
  /** @type {any} */ let value = undefined
  store.subscribe(v => value = v)()
  return value
}
